modules/nw-enforcing-hsts-per-domain.adoc

```diff
@@ -0,0 +1,96 @@
+// Module included in the following assemblies:
+// * networking/configuring-routing.adoc
+
+[id="nw-enforcing-hsts-per-domain_{context}"]
+= Enforcing HTTP strict transport security per-domain
+
+To enforce HTTP strict transport security (HSTS) per-domain for all TLS routes, add a `requiredHSTSPolicies` record to the Ingress spec to capture the configuration of the HSTS policy.
+
+If you configure the `route.openshift.io/RequiredRouteAnnotations` plug-in to enforce HSTS, then any newly created route must be configured with a compliant HSTS policy, which is verified against the global setting on the cluster Ingress configuration, called `ingresses.config.openshift.io/cluster`.
+
+[NOTE]
+====
+To handle upgraded clusters with non-compliant HSTS routes, you can update the manifests at the source and apply the updates.
+====
+
+[IMPORTANT]
+====
+HSTS cannot be applied to non-TLS routes, even if HSTS is requested for all routes globally.
+====
+
+.Prerequisites
+
+* You are logged in to the cluster with a user with `cluster-admin` privileges.
+* You installed the `oc` CLI.
+
+.Procedure
+
+. Edit the Ingress config file:
++
+[source,terminal]
+----
+$ oc edit ingresses.config.openshift.io/cluster
+----
++
+.Example HSTS policy
+[source,yaml]
+----
+apiVersion: config.openshift.io/v1
+kind: Ingress
+metadata:
+  name: cluster
+spec:
+  domain: apps.abc.com
+  requiredHSTSPolicies: <1>
+  - domainPatterns: <2>
+    - abc.com
+    namespaceSelector: <3>
+      matchLabels:
+        myPolicy: strict
+    maxAge: <4>
+      smallestMaxAge: 1
+      largestMaxAge: 31536000
+    preloadPolicy: RequirePreload <5>
+    includeSubDomainsPolicy: RequireIncludeSubDomains
+----
+<1> Required. `requiredHSTSPolicies` are validated in order, and the first matching `domainPatterns` applies.
+<2> Required. You must specify at least one `domainPatterns` hostname.
+<3> Optional. If you include `namespaceSelector`, it must match the labels of the route namespace.
+<4> Required. `max-age` measures the length of time, in seconds, that the HSTS policy is in effect. The client updates `max-age` whenever a response with a HSTS header is received from the host. When `max-age` times out, the client discards the policy.
```

Comment:
```suggestion
<4> Required. `max-age` measures the length of time, in seconds, that the HSTS policy is in effect. This policy setting allows for a smallest and largest `max-age` to be enforced.  
```

Reply:
The `largestMaxAge` value can be left unspecified, in which case no upper limit is enforced, but it must be between 0 and 2147483647.
The `smallestMaxAge` value can be left unspecified, in which case no lower limit is enforced, but it must be between 0 and 2147483647.  Normally choose 0 if you want to allow HSTS to be disabled for troubleshooting, otherwise choose 1 if you never want HSTS to be disabled.